File: app/src/main/java/com/meteomontana/android/data/local/ProfileCache.kt
package com.meteomontana.android.data.local

import android.content.Context
import com.meteomontana.android.domain.model.JournalStats
import com.meteomontana.android.domain.model.PrivateProfile
import com.meteomontana.android.domain.model.SchoolStats
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Caché LOCAL del perfil privado + stats del diario, para poder ver el perfil
 * SIN conexión. Se guarda cada vez que el perfil carga online; offline se lee
 * de aquí. Persiste en SharedPreferences como JSON.
 *
 * Los modelos de dominio no son `@Serializable`, así que usamos snapshots
 * locales (`@Serializable`) y convertimos en ambos sentidos.
 */
class ProfileCache(context: Context) {

    private val prefs = context.getSharedPreferences("profile_cache", Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    fun save(profile: PrivateProfile, stats: JournalStats, followers: Long, following: Long) {
        val snap = Snapshot(
            profile = ProfileSnap(
                profile.uid, profile.email, profile.username, profile.displayName,
                profile.photoUrl, profile.bio, profile.topGrade,
                profile.isPublic, profile.isAdmin, profile.isPremium
            ),
            stats = StatsSnap(
                stats.blockCount, stats.schoolCount, stats.maxGrade,
                stats.bySchool.map { SchoolStatSnap(it.schoolName, it.blockCount, it.maxGrade) },
                stats.boulderCount, stats.routeCount, stats.maxBoulderGrade, stats.maxRouteGrade,
                stats.projectCount, stats.projectBoulderCount, stats.projectRouteCount
            ),
            followers = followers,
            following = following
        )
        prefs.edit().putString(KEY, json.encodeToString(Snapshot.serializer(), snap)).apply()
    }

    /** Último perfil/stats cacheado, o null si nunca cargó online. */
    fun load(): Cached? {
        val raw = prefs.getString(KEY, null) ?: return null
        val snap = runCatching { json.decodeFromString(Snapshot.serializer(), raw) }.getOrNull() ?: return null
        val p = snap.profile
        return Cached(
            profile = PrivateProfile(
                p.uid, p.email, p.username, p.displayName, p.photoUrl, p.bio, p.topGrade,
                p.isPublic, p.isAdmin, p.isPremium
            ),
            stats = JournalStats(
                blockCount = snap.stats.blockCount,
                boulderCount = snap.stats.boulderCount,
                routeCount = snap.stats.routeCount,
                schoolCount = snap.stats.schoolCount,
                maxGrade = snap.stats.maxGrade,
                maxBoulderGrade = snap.stats.maxBoulderGrade,
                maxRouteGrade = snap.stats.maxRouteGrade,
                bySchool = snap.stats.bySchool.map { SchoolStats(it.schoolName, it.blockCount, it.maxGrade) },
                projectCount = snap.stats.projectCount,
                projectBoulderCount = snap.stats.projectBoulderCount,
                projectRouteCount = snap.stats.projectRouteCount
            ),
            followers = snap.followers,
            following = snap.following
        )
    }

    data class Cached(
        val profile: PrivateProfile,
        val stats: JournalStats,
        val followers: Long,
        val following: Long
    )

    @Serializable
    private data class Snapshot(
        val profile: ProfileSnap,
        val stats: StatsSnap,
        val followers: Long,
        val following: Long
    )

    @Serializable
    private data class ProfileSnap(
        val uid: String, val email: String?, val username: String?, val displayName: String?,
        val photoUrl: String?, val bio: String?, val topGrade: String?,
        val isPublic: Boolean, val isAdmin: Boolean, val isPremium: Boolean
    )

    @Serializable
    private data class StatsSnap(
        val blockCount: Int, val schoolCount: Int, val maxGrade: String?,
        val bySchool: List<SchoolStatSnap>,
        val boulderCount: Int = 0, val routeCount: Int = 0,
        val maxBoulderGrade: String? = null, val maxRouteGrade: String? = null,
        val projectCount: Int = 0, val projectBoulderCount: Int = 0, val projectRouteCount: Int = 0
    )

    @Serializable
    private data class SchoolStatSnap(val schoolName: String, val blockCount: Int, val maxGrade: String?)

    private companion object { const val KEY = "profile_snapshot" }
}

File: iosApp/iosApp/DI/ProfileCache.swift
import Foundation
import Shared

/// Caché LOCAL del perfil privado + stats + diario, para poder ver el perfil
/// SIN conexión (espejo del ProfileCache de Android). Se guarda cada vez que el
/// perfil carga online; offline se reconstruye desde aquí. Persiste en
/// UserDefaults como JSON (los modelos Kotlin no son Codable → snapshots propios).
final class ProfileCache {
    static let shared = ProfileCache()
    private let key = "profile_cache_snapshot"
    private init() {}

    func save(profile: PrivateProfile, stats: JournalStats?, entries: [JournalSession], followers: Int64, following: Int64) {
        let snap = Snapshot(
            profile: ProfileSnap(profile),
            stats: stats.map { StatsSnap($0) },
            entries: entries.map { EntrySnap($0) },
            followers: followers,
            following: following
        )
        if let data = try? JSONEncoder().encode(snap) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }

    /// Última copia cacheada reconstruida como modelos Kotlin, o nil si nunca cargó online.
    func load() -> Cached? {
        guard let data = UserDefaults.standard.data(forKey: key),
              let snap = try? JSONDecoder().decode(Snapshot.self, from: data) else { return nil }
        return Cached(
            profile: snap.profile.toModel(),
            stats: snap.stats?.toModel(),
            entries: snap.entries.map { $0.toModel() },
            followers: snap.followers,
            following: snap.following
        )
    }

    struct Cached {
        let profile: PrivateProfile
        let stats: JournalStats?
        let entries: [JournalSession]
        let followers: Int64
        let following: Int64
    }

    // MARK: - Snapshots Codable

    private struct Snapshot: Codable {
        let profile: ProfileSnap
        let stats: StatsSnap?
        let entries: [EntrySnap]
        let followers: Int64
        let following: Int64
    }

    private struct ProfileSnap: Codable {
        let uid: String
        let email: String?, username: String?, displayName: String?
        let photoUrl: String?, bio: String?, topGrade: String?
        let isPublic: Bool, isAdmin: Bool, isPremium: Bool
        init(_ p: PrivateProfile) {
            uid = p.uid; email = p.email; username = p.username; displayName = p.displayName
            photoUrl = p.photoUrl; bio = p.bio; topGrade = p.topGrade
            isPublic = p.isPublic; isAdmin = p.isAdmin; isPremium = p.isPremium
            gender = p.gender
            gearJson = p.gearJson
        }
        var gender: String?
        var gearJson: String?
        func toModel() -> PrivateProfile {
            PrivateProfile(uid: uid, email: email, username: username, displayName: displayName,
                           photoUrl: photoUrl, bio: bio, topGrade: topGrade,
                           isPublic: isPublic, isAdmin: isAdmin, isPremium: isPremium,
                           gender: gender, gearJson: gearJson)
        }
    }

    private struct StatsSnap: Codable {
        let blockCount: Int32, schoolCount: Int32
        let maxGrade: String?
        let bySchool: [SchoolStatSnap]
        let boulderCount: Int32, routeCount: Int32
        let maxBoulderGrade: String?, maxRouteGrade: String?
        let projectCount: Int32, projectBoulderCount: Int32, projectRouteCount: Int32
        init(_ s: JournalStats) {
            blockCount = s.blockCount; schoolCount = s.schoolCount
            maxGrade = s.maxGrade; bySchool = s.bySchool.map { SchoolStatSnap($0) }
            boulderCount = s.boulderCount; routeCount = s.routeCount
            maxBoulderGrade = s.maxBoulderGrade; maxRouteGrade = s.maxRouteGrade
            projectCount = s.projectCount; projectBoulderCount = s.projectBoulderCount
            projectRouteCount = s.projectRouteCount
        }
        func toModel() -> JournalStats {
            JournalStats(blockCount: blockCount, boulderCount: boulderCount, routeCount: routeCount,
                         schoolCount: schoolCount, maxGrade: maxGrade,
                         maxBoulderGrade: maxBoulderGrade, maxRouteGrade: maxRouteGrade,
                         bySchool: bySchool.map { $0.toModel() },
                         projectCount: projectCount, projectBoulderCount: projectBoulderCount,
                         projectRouteCount: projectRouteCount)
        }
    }

    private struct SchoolStatSnap: Codable {
        let schoolName: String, blockCount: Int32, maxGrade: String?
        init(_ s: SchoolStats) { schoolName = s.schoolName; blockCount = s.blockCount; maxGrade = s.maxGrade }
        func toModel() -> SchoolStats {
            SchoolStats(schoolName: schoolName, blockCount: blockCount, maxGrade: maxGrade)
        }
    }

    private struct EntrySnap: Codable {
        let id: String, schoolId: String?, schoolName: String?, sector: String?
        let blockName: String, grade: String?, notes: String?, date: String, createdAt: String
        let discipline: String
        let status: String
        init(_ e: JournalSession) {
            id = e.id; schoolId = e.schoolId; schoolName = e.schoolName; sector = e.sector
            blockName = e.blockName; grade = e.grade; notes = e.notes; date = e.date; createdAt = e.createdAt
            discipline = e.discipline
            status = e.status
        }
        func toModel() -> JournalSession {
            JournalSession(id: id, schoolId: schoolId, schoolName: schoolName, sector: sector,
                           blockName: blockName, grade: grade, notes: notes, date: date, createdAt: createdAt,
                           discipline: discipline, lineId: nil, status: status)
        }
    }
}
